Add `OS` and `OS.POSIX` modules and convert some uses of `Sys`.

gbtitus opened the pull request:
As part of module stabilization we are moving from the existing `Sys` module to an `OS` module with `POSIX` and other submodules, which provide definitions matching the corresponding operating system standards as closely as Chapel can manage.  Here, add the initial `OS.POSIX` module along with tests which exercise (though not exhaustively) what is supported so far. All of the tests pass on both Darwin and Linux. In addition, update some but not all of the module code and tests that used `Sys` to use `OS.POSIX` instead.

The supported POSIX symbols in this initial implementation are the `errno` variable (read-only) along with the following functions and their associated types and constants:
`getenv()`

`strerror()`
`strlen()`

`creat()`
`open()`
`close()`
`read()`
`write()`
`pipe()`

`FD_CLR()`
`FD_ISSET()`
`FD_SET()`
`FD_ZERO()`
`select()`, as `select_posix()` because `select` is a Chapel keyword

`chmod()`
`stat()`

`localtime()`
`localtime_r()`
`asctime()`
`asctime_r()`
`gettimeofday()`
`time()`

gbtitus commented:
For the entire job of converting from `Sys` to `OS.POSIX` I can imagine as many as four PRs, maybe even more if the middle two of the following are split even more finely:
- Add `OS.POSIX` and its tests.
- Convert tests that use `Sys` to use `OS.POSIX` instead.
- Convert modules that use `Sys` to use `OS.POSIX` instead.
- Deprecate `Sys`.

Or, I could imagine combining those into as few as just one PR to do all of it. I'd happily accept feedback on folks' preferences.

bradcray commented:
I don't see any reason to do this all in a single PR, and generally think smaller PRs can be nicer from development / review / testing fallout perspectives.

gbtitus commented:
Smaller would definitely be better from my point of view -- as I said I think I can definitely get step 1 done, and probably some of step 2, and that'll give the group a good place to step off from w.r.t. the rest of the work.

bradcray commented:
Reflecting a bit more, even step 1 could be broken across multiple PRs...

lydia-duncan commented:
I've taken a look at the modules and some of the tests and I think the direction seems reasonable to me so far!

gbtitus commented:
@bradcray do you have a preference for splitting things even more finely, or would you be okay conceptually with this going in as-is, as step 1 in its entirety? (Though of course subject to actual review.)

bradcray commented:
Either is fine with me, I just wanted to note that I don't consider having the entire step 1 done a pre-requisite to starting to PR and merge pieces of it.

gbtitus commented:
Okay, I'm updating some (though not all) of the tests right now, and am leaning toward formally PR'ing this once that's ready.

gbtitus commented:
I've removed the work-in-progress label and updated the description.

mppf commented on modules/standard/OS.chpl:
```diff
@@ -0,0 +1,372 @@
+     There is one unavoidable difference between POSIX and ``OS.POSIX``.
+     POSIX defines a function named ``select()``, which ``OS.POSIX``
+     must define as ``select_posix()`` because ``select`` by itself is
+     a Chapel keyword.
```
well, OS.POSIX can't define `select()` but it could have used a different name.

Suggest 

```
which ``OS.POSIX`` could not use because ``select`` is itself a Chapel keyword.
```

gbtitus replied:
Good point, will do.

mppf commented on modules/standard/OS.chpl:
```diff
@@ -0,0 +1,372 @@
+    //
+    // sys/types.h
+    //
+    extern type blkcnt_t;
```
It would be good to say something about the cast (and other operators) available on the types here; maybe it could be something in the docs comment for POSIX itself or maybe it could be per-type docs comments.

gbtitus replied:
Yep, done: https://github.com/chapel-lang/chapel/pull/19632/commits/ed4153050546e626234ac4f489d5ffe1b2ffffd2.

mppf commented on modules/standard/OS.chpl:
```diff
@@ -0,0 +1,372 @@
+    extern const EXDEV:c_int;
+
+    // POSIX says that errno is a "modifiable lvalue of type int", but
+    // for now we only support reading from it, not assigning to it.
```
Should be a chpldoc comment, maybe with a `.. note:` in it.

gbtitus replied:
Yep, done: https://github.com/chapel-lang/chapel/pull/19632/commits/ef3896558d1285d0c5d1f7e6cc50d5a5c032ba38.

mppf commented on test/modules/OS/POSIX/pipe.chpl:
```diff
@@ -0,0 +1,27 @@
+writeln(ret == 0);
+if ret == 0 {
+  // The pipe was created properly.  Here we create two tasks.  One
+  // writes to the write end and the other reads from the read end.
```
Wouldn't this not work if the two tasks were scheduled on the same thread?

gbtitus replied:
I think at least this one will work, because of the sync var synchronization. But `select.chpl`, below, doesn't have any synchronization.

gbtitus replied:
This is addressed by https://github.com/chapel-lang/chapel/pull/19632/commits/eabbd841b89137ea9b7ade4655241a50e85b5bc7.

mppf commented on test/modules/OS/POSIX/select.chpl:
```diff
@@ -0,0 +1,125 @@
+}
+
+cobegin {
+  beReader(fdsRead);
```
same issue as the pipes test - if there was only 1 thread, we would have a problem here, right?

gbtitus replied:
This is addressed by https://github.com/chapel-lang/chapel/pull/19632/commits/eabbd841b89137ea9b7ade4655241a50e85b5bc7.

mppf reviewed (approved):
Thanks @gbtitus !

gbtitus commented:
@mppf I think I should limit the configurations in which we test this module, similar to what `test/users/ferguson/sys/getenv.skipif` used to do. If I did that and added `CHPL_TASKS != qthreads` and added sync var synchronization to `select.chpl` would that get us everything we need there?

gbtitus commented:
I've confirmed with @ronawho that as long as the tasks do cooperative yields (including implicitly by means of sync vars), at least with qthreads tasking both tasks can share a single thread without CPU starvation and deadlock. So, I'm going to proceed in that direction.

mppf commented on test/modules/OS/POSIX/pipe.chpl:
```diff
@@ -0,0 +1,27 @@
+use OS.POSIX;
+
+var fildes:2*c_int;
```
I don't think we always guarantee that tuple variables are stored contiguously in memory. So it would be better to use c_array here to be more future-proof. I do expect this will work as a global variable like here.

gbtitus replied:
Hmm. I'd used a tuple because of the advice in https://chapel-lang.org/docs/language/spec/interoperability.html#referring-to-external-c-types. But it would be easy to change this to use `c_array` instead. What's your preference?

mppf replied:
My preference would be to use a `c_array` but I wouldn't view it as a show-stopper here by any means.

gbtitus replied:
Agreed, module and test changes done in https://github.com/chapel-lang/chapel/pull/19632/commits/8138780da7db75b6e8c0471f8aaa108fdea63a7c.

bradcray commented:
Thanks so much for taking this on, Greg!  I hadn't been expecting you to own this (or to need to), but it makes for a great parting gift, and I appreciate the effort!
